# CS 6220 HW3 -- Xinan Wang

In [22]:
import pandas as pd
import numpy as np 
import matplotlib.pyplot as plt
import seaborn as sns
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

## 1 Online Retail Market Basket Analysis [50 points]

In [2]:
import pandas as pd
pd.set_option("max_colwidth", 150)
f = "https://github.com/cs6220/cs6220.spring2019/raw/master/data/Online%20Retail.xlsx"
df = pd.read_excel(f)
basket = (df[df["Country"] == "United Kingdom"]
        .groupby(["InvoiceNo", "Description"])["Quantity"]
        .sum().unstack().reset_index().fillna(0)
        .set_index("InvoiceNo")) # transform transactions into baskets of items
basket_sets = basket.applymap(lambda x: 1 if x >=1 else 0) # convert counts to booleans

In [3]:
df.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom

In [4]:
basket_sets

Description  20713   4 PURPLE FLOCK DINNER CANDLES  \
InvoiceNo                                            
536365           0                               0   
536366           0                               0   
536367           0                               0   
536368           0                               0   
536369           0                               0   
...            ...                             ...   
C581484          0                               0   
C581490          0                               0   
C581499          0                               0   
C581568          0                               0   
C581569          0                               0   

Description   50'S CHRISTMAS GIFT BAG LARGE   DOLLY GIRL BEAKER  \
InvoiceNo                                                         
536365                                    0                   0   
536366                                    0                   0   
536367                                    0                   0   
536368                                    0                   0   
536369                                    0                   0   
...                                     ...                 ...   
C581484                                   0                   0   
C581490                                   0                   0   
C581499                                   0                   0   
C581568                                   0                   0   
C581569                                   0                   0   

Description   I LOVE LONDON MINI BACKPACK   NINE DRAWER OFFICE TIDY  \
InvoiceNo                                                             
536365                                  0                         0   
536366                                  0                         0   
536367                                  0                         0   
536368                                  0                         0   
536369                                  0                         0   
...                                   ...                       ...   
C581484                                 0                         0   
C581490                                 0                         0   
C581499                                 0                         0   
C581568                                 0                         0   
C581569                                 0                         0   

Description   OVAL WALL MIRROR DIAMANTE    RED SPOT GIFT BAG LARGE  \
InvoiceNo                                                            
536365                                 0                         0   
536366                                 0                         0   
536367                                 0                         0   
536368                                 0                         0   
536369                                 0                         0   
...                                  ...                       ...   
C581484                                0                         0   
C581490                                0                         0   
C581499                                0                         0   
C581568                                0                         0   
C581569                                0                         0   

Description   SET 2 TEA TOWELS I LOVE LONDON    SPACEBOY BABY GIFT SET  ...  \
InvoiceNo                                                               ...   
536365                                      0                        0  ...   
536366                                      0                        0  ...   
536367                                      0                        0  ...   
536368                                      0                        0  ...   
536369                                      0                        0  ...   
...                                  

### 1.1 Frequent Itemset Generation

Use the Apriori algorithm to generate frequent itemsets with up to 2 items per set. Explore dif-
ferent support thresholds until you obtain at least 5 results for both 1- and 2-itemsets. Note that
mlxtend offers an implementation of the Apriori algorithm (see mlxtend.frequent patterns).

#### [10 pts] What are the top 5 1-itemsets with the highest support?

In [13]:
# Generate the frequent itemset and find the top 5 1-itemset support

frequent_itemsets = apriori(basket_sets, min_support = 0.05, use_colnames = True)
frequent_itemsets = frequent_itemsets.sort_values("support", ascending = False)
top_5_1item = frequent_itemsets[frequent_itemsets['itemsets'].apply(len) == 1][:5]
top_5_1item

/Users/nannmemeda/opt/anaconda3/lib/python3.9/site-packages/mlxtend/frequent_patterns/fpcommon.py:111: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


support                              itemsets
14  0.098276  (WHITE HANGING HEART T-LIGHT HOLDER)
3   0.087931             (JUMBO BAG RED RETROSPOT)
12  0.076452            (REGENCY CAKESTAND 3 TIER)
11  0.072323                       (PARTY BUNTING)
7   0.063158             (LUNCH BAG RED RETROSPOT)

#### [10 pts] What are the top 5 2-itemsets with the highest support?

In [10]:
# Find the top 5 2 itemsets with the highest support

frequent_itemsets = apriori(basket_sets, min_support=0.02, use_colnames=True)
frequent_itemsets = frequent_itemsets.sort_values(by="support", ascending=False)
top_5_2items = frequent_itemsets[frequent_itemsets['itemsets'].apply(len) == 2][:5]
top_5_2items

/Users/nannmemeda/opt/anaconda3/lib/python3.9/site-packages/mlxtend/frequent_patterns/fpcommon.py:111: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


support  \
211  0.035617   
207  0.031806   
218  0.031670   
217  0.029809   
223  0.027541   

                                                                itemsets  
211                   (JUMBO BAG PINK POLKADOT, JUMBO BAG RED RETROSPOT)  
207  (GREEN REGENCY TEACUP AND SAUCER, ROSES REGENCY TEACUP AND SAUCER )  
218                    (JUMBO BAG RED RETROSPOT, JUMBO STORAGE BAG SUKI)  
217         (JUMBO SHOPPER VINTAGE RED PAISLEY, JUMBO BAG RED RETROSPOT)  
223                   (LUNCH BAG RED RETROSPOT, LUNCH BAG  BLACK SKULL.)

#### [5 pts] What is the highest support value for the 1-itemsets?

In [20]:
# Find the highest support value for 1-itemsets

frequent_itemsets = apriori(basket_sets, min_support = 0.05, use_colnames = True)
frequent_itemsets = frequent_itemsets.sort_values("support", ascending = False)
highest_support_1itemset = frequent_itemsets[frequent_itemsets['itemsets'].apply(len) == 1]['support'].values[0]
print("The highest support value for 1-itemsets is: ",highest_support_1itemset)

/Users/nannmemeda/opt/anaconda3/lib/python3.9/site-packages/mlxtend/frequent_patterns/fpcommon.py:111: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


The highest support value for 1-itemsets is:  0.09827586206896552


#### [5 pts] What is the highest support value for the 2-itemsets?

In [21]:
# Find the highest support value for 2-itemsets

frequent_itemsets = apriori(basket_sets, min_support = 0.02, use_colnames = True)
frequent_itemsets = frequent_itemsets.sort_values("support", ascending = False)
highest_support_2itemset = frequent_itemsets[frequent_itemsets['itemsets'].apply(len) == 2]['support'].values[0]
print("The highest support value for 2-itemsets is: ",highest_support_2itemset)

/Users/nannmemeda/opt/anaconda3/lib/python3.9/site-packages/mlxtend/frequent_patterns/fpcommon.py:111: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


The highest support value for 2-itemsets is:  0.03561705989110708


### 1.2 Association Rule Generation

Generate association rules from the frequent itemsets. Use confidence as the metric to evaluate the association rules. Explore different confidence thresholds until you obtain at least 5 association rules. Note that mlxtend may be used to generate association rules from frequent itemsets (see mlxtend.frequent patterns).

#### [10 pts] What are the top 5 association rules?

In [23]:
rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.1)
top_5_rules = rules.sort_values(by = 'confidence', ascending = False)[:5]
top_5_rules

antecedents  \
50  (PINK REGENCY TEACUP AND SAUCER, ROSES REGENCY TEACUP AND SAUCER )   
48   (GREEN REGENCY TEACUP AND SAUCER, PINK REGENCY TEACUP AND SAUCER)   
11                                    (PINK REGENCY TEACUP AND SAUCER)   
18                                    (PINK REGENCY TEACUP AND SAUCER)   
2                                    (GREEN REGENCY TEACUP AND SAUCER)   

                           consequents  antecedent support  \
50   (GREEN REGENCY TEACUP AND SAUCER)            0.024773   
48  (ROSES REGENCY TEACUP AND SAUCER )            0.026180   
11   (GREEN REGENCY TEACUP AND SAUCER)            0.031897   
18  (ROSES REGENCY TEACUP AND SAUCER )            0.031897   
2   (ROSES REGENCY TEACUP AND SAUCER )            0.042377   

    consequent support   support  confidence       lift  leverage  conviction  
50            0.042377  0.022368    0.902930  21.306837  0.021319    9.865319  
48            0.043421  0.022368    0.854419  19.677538  0.021232    6.570786  
11            0.042377  0.026180    0.820768  19.368019  0.024828    5.342926  
18            0.043421  0.024773    0.776671  17.886978  0.023388    4.283280  
2             0.043421  0.031806    0.750535  17.285056  0.029966    3.834527

#### [10 pts] What items make up one of the top association rules? Search online for the items (or at least items with the same name). Do you think they are likely to be bought together?

According to the previous question answer, we can see that the top association rule with the highest confidence is pink regency teacup and saucer, roses regency teacup and saucer => green regency teacup and saucer, which means the person who purchase pink regency teacup and saucer and roses regency teacup and saucer have large possibility to buy green regency teacup and saucer as well. In my personal opinion, I think they are very likely to be bought together. Because when a person buy 2 color teacup and saucers, they probably also purchase the third color because they need more different color teacups and saucers. Also, different colors teacup and saucers are very likely to be bought together because when a person decorate their new home, they may need to buy a lot of teacups and saucers together. So, to conclude, I think they are very likely to be bought together.

## 2 Association Rule Mining U.S. Census Data [50 points]

Load the Adult dataset from here. The dataset is an extraction from the 1994 U.S. Census.

More information on the dataset can be found here

Below is a snippet of code that you can use to load that dataset into a pandas DataFrame

In [41]:
import numpy as np
import pandas as pd
path = "https://raw.githubusercontent.com/cs6220/cs6220.spring2019/master/data/adult/"
names = pd.read_table(path + "adult.names", delimiter=None, header=None)
parse_cols = lambda x: x.str.split(":", expand=True).iloc[:, 0]
columns = np.roll(parse_cols(names.iloc[92:108, 0]), shift=-1)
df_adult = pd.read_csv(path + "adult.data", delimiter=None, header=None, index_col=False)
df_adult.columns = columns

Your task is to perform association rule mining on the dataset and identify interesting rules.

### 2.1 Association Rule Mining

Transform the data and perform association rule mining on the dataset. Note that one-hot encoding can be performed using pandas (see pandas.get dummies), patsy (see Coding categorial data with patsy), or scikit-learn (see sklearn.preprocessing.OneHotEncoder). You are welcome to use whichever option you prefer.

#### [20 pts] Transform the raw dataset into a format appropriate for association rule mining by dropping all continuous columns and one-hot encoding the remaining columns. The values for each resulting column should be binary, represented by a 1 or 0.

In [42]:
df_adult.head()

age          workclass  fnlwgt   education  education-num  \
0   39          State-gov   77516   Bachelors             13   
1   50   Self-emp-not-inc   83311   Bachelors             13   
2   38            Private  215646     HS-grad              9   
3   53            Private  234721        11th              7   
4   28            Private  338409   Bachelors             13   

        marital-status          occupation    relationship    race      sex  \
0        Never-married        Adm-clerical   Not-in-family   White     Male   
1   Married-civ-spouse     Exec-managerial         Husband   White     Male   
2             Divorced   Handlers-cleaners   Not-in-family   White     Male   
3   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male   
4   Married-civ-spouse      Prof-specialty            Wife   Black   Female   

   capital-gain  capital-loss  hours-per-week  native-country >50K, <=50K.  
0          2174             0              40   United-States        <=50K  
1             0             0              13   United-States        <=50K  
2             0             0              40   United-States        <=50K  
3             0             0              40   United-States        <=50K  
4             0             0              40            Cuba        <=50K

In [43]:
df_adult = df_adult.drop(df_adult.columns[(df_adult.dtypes == float) | (df_adult.dtypes == int)], axis = 1)
df_adult.head()

workclass   education       marital-status          occupation  \
0          State-gov   Bachelors        Never-married        Adm-clerical   
1   Self-emp-not-inc   Bachelors   Married-civ-spouse     Exec-managerial   
2            Private     HS-grad             Divorced   Handlers-cleaners   
3            Private        11th   Married-civ-spouse   Handlers-cleaners   
4            Private   Bachelors   Married-civ-spouse      Prof-specialty   

     relationship    race      sex  native-country >50K, <=50K.  
0   Not-in-family   White     Male   United-States        <=50K  
1         Husband   White     Male   United-States        <=50K  
2   Not-in-family   White     Male   United-States        <=50K  
3         Husband   Black     Male   United-States        <=50K  
4            Wife   Black   Female            Cuba        <=50K

In [44]:
df_adult = pd.get_dummies(df_adult, drop_first = True)
df_adult.head()

workclass_ Federal-gov  workclass_ Local-gov  workclass_ Never-worked  \
0                       0                     0                        0   
1                       0                     0                        0   
2                       0                     0                        0   
3                       0                     0                        0   
4                       0                     0                        0   

   workclass_ Private  workclass_ Self-emp-inc  workclass_ Self-emp-not-inc  \
0                   0                        0                            0   
1                   0                        0                            1   
2                   1                        0                            0   
3                   1                        0                            0   
4                   1                        0                            0   

   workclass_ State-gov  workclass_ Without-pay  education_ 11th  \
0                     1                       0                0   
1                     0                       0                0   
2                     0                       0                0   
3                     0                       0                1   
4                     0                       0                0   

   education_ 12th  ...  native-country_ Puerto-Rico  \
0                0  ...                            0   
1                0  ...                            0   
2                0  ...                            0   
3                0  ...                            0   
4                0  ...                            0   

   native-country_ Scotland  native-country_ South  native-country_ Taiwan  \
0                         0                      0                       0   
1                         0                      0                       0   
2                         0                      0                       0   
3                         0                      0                       0   
4                         0                      0                       0   

   native-country_ Thailand  native-country_ Trinadad&Tobago  \
0                         0                                0   
1                         0                                0   
2                         0                                0   
3                         0                                0   
4                         0                                0   

   native-country_ United-States  native-country_ Vietnam  \
0                              1                        0   
1                              1                        0   
2                              1                        0   
3                              1                        0   
4                              0                        0   

   native-country_ Yugoslavia  >50K, <=50K._ >50K  
0                           0                   0  
1                           0                   0  
2                           0                   0  
3                           0                   0  
4                           0                   0  

[5 rows x 95 columns]

#### [10 pts] Use confidence for the rule interestingness (metric="confidence") and generate rules up to a depth of 3 (max len=3). Generate rules and flag at least 5 examples that you find interesting. Comment on your findings and share a few notes on what you learned from the rules you are highlighting. Note that you can choose your support and confidence thresholds yourself in this analysis.

In [61]:
frequent_itemsets = apriori(df_adult, min_support=0.2, max_len=3, use_colnames=True)
rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.1)
top_5_confidence_rules = rules.sort_values(by = 'confidence', ascending = False)[:5]
top_5_confidence_rules

/Users/nannmemeda/opt/anaconda3/lib/python3.9/site-packages/mlxtend/frequent_patterns/fpcommon.py:111: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


antecedents  \
158                    (race_ White, >50K, <=50K._ >50K)   
110                    (race_ White, education_ HS-grad)   
157  (native-country_ United-States, >50K, <=50K._ >50K)   
144           (race_ White, relationship_ Not-in-family)   
116                      (sex_ Male, education_ HS-grad)   

                         consequents  antecedent support  consequent support  \
158  (native-country_ United-States)            0.218574            0.895857   
110  (native-country_ United-States)            0.273456            0.895857   
157                    (race_ White)            0.220233            0.854274   
144  (native-country_ United-States)            0.218820            0.895857   
116  (native-country_ United-States)            0.218390            0.895857   

      support  confidence      lift  leverage  conviction  
158  0.205890    0.941970  1.051474  0.010079    1.794638  
110  0.257486    0.941599  1.051060  0.012508    1.783248  
157  0.205890    0.934877  1.094353  0.017751    2.237697  
144  0.202973    0.927579  1.035410  0.006941    1.438021  
116  0.202543    0.927436  1.035250  0.006897    1.435195

I have generated the rules with top 5 highest confidence. The higher the confidence, the more likely it is for the consequents to be present in transactions containing antecedents.

The first rule is (race white & salary > 50K) => (native country is United States), and it has 0.941970 confidence. This means we have approximately 94.20% probability that the person native country is United States when the person is white race and salary above 50K. This is a very high probability.

The second rule is (race white & education is high school graduate) => (native country is United States), and it has 0.941599 confidence. This means we have approximately 94.16% probability that the person native country is United States when the person is white race and education level is high school graduate. This is a very high probability.

The third rule is (native country is United States & salary > 50K) => (race white), and it has 0.934877 confidence. This means we have approximately 93.49% probability that the person race is white when the person native country is United States and his/her salary is above 50K. This is a very high probability.

The first rule is (race white & relationship is not in family) => (native country is United States), and it has 0.927579 confidence. This means we have approximately 92.76% probability that the person native country is United States when the person is white race and his/her relationship is not in family. This is a very high probability.

The final rule is (sex is Male & education is high school graduate) => (native country is United States), and it has 0.927436 confidence. This means we have approximately 92.74% probability that the person native country is United States when the person is a Male and his education level is high school graduate. This is a very high probability.

#### [10 pts] Use lift for the rule interestingness (metric="lift") and generate rules up to a depth of 3 (max len=3). Generate rules and flag at least 5 examples that you find interesting. Comment on your findings and share a few notes on what you learned from the rules you are highlighting. Note that you can choose your support and confidence thresholds yourself in this analysis.

In [64]:
frequent_itemsets = apriori(df_adult, min_support=0.2, max_len=3, use_colnames=True)
rules = association_rules(frequent_itemsets, metric="lift", min_threshold=0.1)
top_5_lift_rules = rules.sort_values(by = 'lift', ascending = False)[:10]
top_5_lift_rules

/Users/nannmemeda/opt/anaconda3/lib/python3.9/site-packages/mlxtend/frequent_patterns/fpcommon.py:111: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


antecedents  \
27                                                  (>50K, <=50K._ >50K)   
26                                  (marital-status_ Married-civ-spouse)   
125                                 (marital-status_ Married-civ-spouse)   
120                                             (race_ White, sex_ Male)   
121                    (race_ White, marital-status_ Married-civ-spouse)   
124                                                          (sex_ Male)   
136                                                          (sex_ Male)   
133  (native-country_ United-States, marital-status_ Married-civ-spouse)   
137                                 (marital-status_ Married-civ-spouse)   
132                           (native-country_ United-States, sex_ Male)   

                                                             consequents  \
27                                  (marital-status_ Married-civ-spouse)   
26                                                  (>50K, <=50K._ >50K)   
125                                             (race_ White, sex_ Male)   
120                                 (marital-status_ Married-civ-spouse)   
121                                                          (sex_ Male)   
124                    (race_ White, marital-status_ Married-civ-spouse)   
136  (native-country_ United-States, marital-status_ Married-civ-spouse)   
133                                                          (sex_ Male)   
137                           (native-country_ United-States, sex_ Male)   
132                                 (marital-status_ Married-civ-spouse)   

     antecedent support  consequent support   support  confidence      lift  \
27             0.240810            0.459937  0.205522    0.853463  1.855609   
26             0.459937            0.240810  0.205522    0.446848  1.855609   
125            0.459937            0.588864  0.369645    0.803686  1.364807   
120            0.588864            0.459937  0.369645    0.627725  1.364807   
121            0.411842            0.669205  0.369645    0.897539  1.341201   
124            0.669205            0.411842  0.369645    0.552363  1.341201   
136            0.669205            0.410553  0.366911    0.548279  1.335466   
133            0.410553            0.669205  0.366911    0.893701  1.335466   
137            0.459937            0.598507  0.366911    0.797743  1.332888   
132            0.598507            0.459937  0.366911    0.613044  1.332888   

     leverage  conviction  
27   0.094765    3.685497  
26   0.094765    1.372481  
125  0.098804    2.094277  
120  0.098804    1.450711  
121  0.094037    3.228497  
124  0.094037    1.313918  
136  0.092167    1.304893  
133  0.092167    3.111936  
137  0.091636    1.985062  
132  0.091636    1.395671

I have generated the rules with top 5 highest lift. Lift is calculated as the ratio of the observed support for  given itemset to the expected support if the items were independent. A lift value greater than 1 indicates that the items are more likely to appear together in transactions than if they were independent.

The first rule is (salary > 50K) => (marital status is Married-civ-spouse), and it has 1.855609 lift. This means when people with annual income greater than $50,000 have approximately 1.86 times more likely to be married  and have a civil spouse, compared to their income and marital status were independent.

The second rule is (marital-status_ Married-civ-spouse) => (race_ White, sex_ Male), and it has 1.364807 lift. This means when people are married and have a civil spouse have approximately 1.36 times more likely to be white race and a male, compared to their race, sex and marital status were independent.

The third rule is (race_ White, marital-status_ Married-civ-spouse) => (native-country_ United-States, marital-status_ Married-civ-spouse), and it has 1.341201 lift. This means when people have a civil spouse, and his/her race is white, will have approximately 1.34 times more likely to be a male, compared to their race, sex and marital status were independent.

The forth rule is (native-country_ United-States, marital-status_ Married-civ-spouse) => (native-country_ United-States, sex_ Male), and it has 1.332888 lift. This means when people are married and have a civil spouse, and his/her native country is United States, will have approximately 1.34 times more likely to be a male, compared to their race, native country and marital status were independent.

The final rule is (marital-status_ Married-civ-spouse) => (native-country_ United-States, sex_ Male), and it has 1.332888 lift. This means when people are married and have a civil spouse will have approximately 1.34 times more likely to be a male,and his native country is United States, compared to their race, native country and marital status were independent.

#### [10 pts] Compare the top rules using the two interestingness measures for the same levels of support (use at least two different levels of support) and comment on your findings.

In [68]:
# Case when min_support is 0.1

frequent_itemsets_1 = apriori(df_adult, min_support = 0.1, use_colnames = True)
rule_1_confidence = association_rules(frequent_itemsets_1, metric = 'confidence', min_threshold = 0.7)
rule_1_lift = association_rules(frequent_itemsets_1, metric = 'lift', min_threshold = 1.2)

/Users/nannmemeda/opt/anaconda3/lib/python3.9/site-packages/mlxtend/frequent_patterns/fpcommon.py:111: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


In [73]:
print("Confidence:")
rule_1_confidence.sort_values(by='confidence', ascending=False)[:3]

Confidence:


antecedents  \
126                                                                                 (race_ White, occupation_ Craft-repair)   
321                                        (race_ White, sex_ Male, marital-status_ Married-civ-spouse, education_ HS-grad)   
345  (>50K, <=50K._ >50K, sex_ Male, workclass_ Private, native-country_ United-States, marital-status_ Married-civ-spouse)   

                         consequents  antecedent support  consequent support  \
126                      (sex_ Male)            0.113449            0.669205   
321  (native-country_ United-States)            0.119652            0.895857   
345                    (race_ White)            0.106047            0.854274   

      support  confidence      lift  leverage  conviction  
126  0.108228    0.953979  1.425540  0.032307    7.187970  
321  0.113694    0.950205  1.060666  0.006503    2.091449  
345  0.100765    0.950188  1.112276  0.010171    2.925544

In [74]:
print("Lift:")
rule_1_lift.sort_values(by='lift', ascending=False)[:3]

Lift:


antecedents  \
213                                       (race_ White, relationship_ Own-child)   
210               (marital-status_ Never-married, native-country_ United-States)   
208  (marital-status_ Never-married, native-country_ United-States, race_ White)   

                                                        consequents  \
213  (marital-status_ Never-married, native-country_ United-States)   
210                          (race_ White, relationship_ Own-child)   
208                                       (relationship_ Own-child)   

     antecedent support  consequent support  support  confidence      lift  \
213            0.130678            0.294186  0.11004    0.842068  2.862364   
210            0.294186            0.130678  0.11004    0.374047  2.862364   
208            0.247658            0.155646  0.11004    0.444320  2.854680   

     leverage  conviction  
213  0.071596    4.469103  
210  0.071596    1.388799  
208  0.071493    1.519497

Compared the same levels support with different measures(confidence and lift) and same level support, we can find that it will generate different top rules when we use different measures. 

For example, the top 1 rule using confidence measure is (race_ White, occupation_ Craft-repair) => (sex_ Male), and its confidence is 0.953979, lift is 1.425540. These means we have approximately 95.40% probability that the is a male when the person's race is white and occupation is craft-repair. Also, when people race is white and occupation is craft repair, it has approximately 1.43 times more likely to be a male.

However, the top 1 rule using lift measure is (race_ White, relationship_ Own-child) => (marital-status_ Never-married, native-country_ United-States), and its confidence is 0.842068, lift is 2.862364. Thease means we have approximately 84.21% probability that the person's marital status is never married and his/her native country is United States when the person's race is white and relationship is own-child. Also, when people race is white and relationship is own-child, it has approximately 2.86 times more likely that their marital status is never married and his/her native country is United States.

Therefore, the different measures will generate different top rules. We should choose the measures by needs.

In [76]:
# Case when min_support is 0.05

frequent_itemsets_2 = apriori(df_adult, min_support = 0.05, use_colnames = True)
rule_2_confidence = association_rules(frequent_itemsets_2, metric = 'confidence', min_threshold = 0.7)
rule_2_lift = association_rules(frequent_itemsets_2, metric = 'lift', min_threshold = 1.2)

/Users/nannmemeda/opt/anaconda3/lib/python3.9/site-packages/mlxtend/frequent_patterns/fpcommon.py:111: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


In [78]:
print("Confidence:")
rule_2_confidence.sort_values(by='confidence', ascending=False)[:3]

Confidence:


antecedents  \
795  (race_ White, marital-status_ Married-civ-spouse, occupation_ Craft-repair, native-country_ United-States)   
531                                 (occupation_ Craft-repair, marital-status_ Married-civ-spouse, race_ White)   
537               (occupation_ Craft-repair, marital-status_ Married-civ-spouse, native-country_ United-States)   

     consequents  antecedent support  consequent support   support  \
795  (sex_ Male)            0.067412            0.669205  0.066828   
531  (sex_ Male)            0.072940            0.669205  0.072295   
537  (sex_ Male)            0.071343            0.669205  0.070667   

     confidence      lift  leverage  conviction  
795    0.991344  1.481375  0.021716   38.215471  
531    0.991158  1.481096  0.023483   37.411284  
537    0.990529  1.480157  0.022924   34.928893

In [79]:
print("Lift:")
rule_2_lift.sort_values(by='lift', ascending=False)[:3]

Lift:


antecedents  \
1275                                     (race_ White, relationship_ Own-child)   
1272  (marital-status_ Never-married, native-country_ United-States, sex_ Male)   
476                  (marital-status_ Married-civ-spouse, education_ Bachelors)   

                                                                    consequents  \
1275  (marital-status_ Never-married, native-country_ United-States, sex_ Male)   
1272                                     (race_ White, relationship_ Own-child)   
476                                             (sex_ Male, >50K, <=50K._ >50K)   

      antecedent support  consequent support   support  confidence      lift  \
1275            0.130678            0.161267  0.062283    0.476616  2.955453   
1272            0.161267            0.130678  0.062283    0.386212  2.955453   
476             0.085010            0.204601  0.050674    0.596098  2.913473   

      leverage  conviction  
1275  0.041209    1.602519  
1272  0.041209    1.416324  
476   0.033281    1.969289

With different min_support value, we can see that the transactions and the top itemsets has totally changed. For example, when min_support is 0.1, under confidence measure, the top rule is (race_ White, occupation_ Craft-repair) => (sex_ Male). However, when min_support is 0.05, the top rule is (race_ White, marital-status_ Married-civ-spouse, occupation_ Craft-repair, native-country_ United-States) => (sex_ Male). And their confidence or lift or other measures are increased as well by the min_support decreasing. For example, under confidence measure and min_support is 0.1, the top rule's confidence is 0.953979, while it becomes 0.991344 when the the min_support decrease to 0.05.

Therefore, different min_support value will generate different quantity itemsets, and may influence the itemsets that we choose in the final decision. The purpose of setting a minimum support threshold is to reduce the number of candidate itemsets that need to be considered for rule generation, as well as to control the size of the rules that are generated. If the minimum support threshold is set too low, then many meaningless or uninteresting itemsets will be generated. If the threshold is set too high, then very few itemsets will be generated and many interesting patterns may be missed. 